stablehlo/transforms/StablehloLegalizeToVhlo.cpp

```diff
@@ -182,261 +188,377 @@ Attribute convertAttrToVhlo(Attribute stablehloAttr,
       vhloIndexVectorDim);
-  return success();
+  return specialSuccess();
 }
 
-LogicalResult convertScatterDimensionNumbers(
+SpecialResult convertScatterDimensionNumbers(
     const ConversionPattern& pattern, Attribute stablehloAttr,
     SmallVector<NamedAttribute>& vhloAttrs) {
   auto attr = stablehloAttr.dyn_cast<stablehlo::ScatterDimensionNumbersAttr>();
-  if (!attr) return failure();
+  if (!attr) return specialFailure();
 
   auto vhloUpdateWindowDims = convertInts(pattern, attr.getUpdateWindowDims());
-  if (!vhloUpdateWindowDims) return failure();
+  if (!vhloUpdateWindowDims) return specialFailure();
   vhloAttrs.emplace_back(
       StringAttr::get(pattern.getContext(), "update_window_dims"),
       vhloUpdateWindowDims);
 
   auto vhloInsertedWindowDims =
       convertInts(pattern, attr.getInsertedWindowDims());
-  if (!vhloInsertedWindowDims) return failure();
+  if (!vhloInsertedWindowDims) return specialFailure();
   vhloAttrs.emplace_back(
       StringAttr::get(pattern.getContext(), "inserted_window_dims"),
       vhloInsertedWindowDims);
 
   auto vhloScatterDimsToOperandDims =
       convertInts(pattern, attr.getScatterDimsToOperandDims());
-  if (!vhloScatterDimsToOperandDims) return failure();
+  if (!vhloScatterDimsToOperandDims) return specialFailure();
   vhloAttrs.emplace_back(
       StringAttr::get(pattern.getContext(), "scatter_dims_to_operand_dims"),
       vhloScatterDimsToOperandDims);
 
   auto vhloIndexVectorDim = convertInt(pattern, attr.getIndexVectorDim());
-  if (!vhloIndexVectorDim) return failure();
+  if (!vhloIndexVectorDim) return specialFailure();
   vhloAttrs.emplace_back(
       StringAttr::get(pattern.getContext(), "index_vector_dim"),
       vhloIndexVectorDim);
-  return success();
+  return specialSuccess();
+}
+
+SpecialResult convertUseGlobalDeviceIds(
+    const ConversionPattern& pattern, Attribute stablehloAttr,
+    SmallVector<NamedAttribute>& vhloAttrs) {
+  if (!stablehloAttr.isa<UnitAttr>()) return specialFailure();
+  vhloAttrs.emplace_back(
+      StringAttr::get(pattern.getContext(), "use_global_device_ids"),
+      vhlo::BooleanV1Attr::get(pattern.getContext(), true));
+  return specialSuccess();
+}
+
+template <typename StablehloOpTy>
+SpecialResult convertSpecial(const OpConversionPattern<StablehloOpTy>& pattern,
+                             StringRef stablehloName, Attribute stablehloAttr,
+                             SmallVector<NamedAttribute>& vhloAttrs) {
+  if constexpr (std::is_same<StablehloOpTy, stablehlo::AllGatherOp>::value ||
+                std::is_same<StablehloOpTy, stablehlo::AllReduceOp>::value ||
+                std::is_same<StablehloOpTy, stablehlo::AllToAllOp>::value ||
+                std::is_same<StablehloOpTy,
+                             stablehlo::CollectivePermuteOp>::value ||
+                std::is_same<StablehloOpTy,
+                             stablehlo::ReduceScatterOp>::value) {
+    if (stablehloName == "channel_handle") {
```

Comment:
How about no curly braces on the `if (stablehloName == "...")` but keep them on the constexpr ones for dangling if protection

Reply:
Done.